### DISCUSSION TITLE
Calculation/Presentation of Maximum Value

### DISCUSSION THREAD

**Discussion by wxivd** [asastats/channel#110] [Ideas]
I think the current algorithms prioritise tinyman.org. and then introduce other AMM to offer maximum value. The total value from both AMMs adds to the total value presented (Balance Box). 

I discuss the following aspects:

1. The maximum value mainly solve the problem of low liquidity in one particular pool.
- This problem is expected to solve itself as the blockchain grows. E.g. Deeper liquidity is provided. Therefor the feature may end up being absolute. 

2. Prioritizing one AMM over another could be a potential service we provide to AMMs in the future. E.g. Featuring as a first option in the way we present/order it.

3. Currently, presenting favourable exchange value from the second AMM looks short of ideal. See the example below. 
<img width="1150" alt="Screenshot 2022-05-26 at 20 40 37" src="https://user-images.githubusercontent.com/95645978/170564753-84052f15-8192-4bdd-8558-cc0408379e54.png">
<img width="423" alt="Screenshot 2022-05-26 at 20 40 47" src="https://user-images.githubusercontent.com/95645978/170564781-7548aef7-83b8-46e5-96d6-0b3396161ff1.png">

I have the following points
- Exchange value on Tinyman is higher than is shown
- ASA amount to be exchanged on second/ third AMMs are getting smaller and smaller which is solid math for the current method of calculation. At times the amount becomes too low that it's kind of counterproductive to exchange < 2 Algo worth of Tokens on the exchange. 

I suggest the following modification:
1. At all time, present the maximum amount of a given ASA that an investor own and can exchange on available exchanges for maximum value.
2. We control the ordering of AMMs either as a feature for AMMs or based on maximum amount of ASA to be sold at maximum value (Tinyman is currently leading the way).
3. This method has a drawback that adding up amounts from all AMMs does not reflect the maximum value of asset held in wallet as is presented under "Balance". I dont see it as a problem really.

**Comment by wxivd** [comment]
> Exchange value on Tinyman is higher than is shown

I correct myself. There is no error here. I mixed up exchanging maximum amount and maximum return on value. Exchange the maximum amount would adversely affect the value. 

This leaves my previous comment as a suggestion only regarding points 1, 2 & 3.

**Comment by ipaleka** [comment]
I'm not exactly sure what you mean by ordering (EDIT: I've tried to reformulate this beginning - something like "I can't comprehend how we can justify usage of a different ordering and explain that to our users") and also the engine doesn't have a mechanism for prioritizing besides math itself.

AMMs are ordered by the amount of Algo, or to say the same thing, by the asset amount user should sell on the particular AMM.

The actual price row exists solely for purpose of identifying imperfections of the system used. Right now the input parameters for our engine are `(2, 2, 10)` (let's skip their meaning for now). If we change those parameters to `(0.01, 0.01, 90)` the engine would make around 1000 times more calculations per bundle/address and **all the AMMs** would get exactly the same price (up to the ninth decimal), let's say it would become:

`0.39500107 ALGO/comand`

This system has no alternative as there can exist only one maximum.

P.S. our slippage is 0

**Reply by wxivd** [reply]
> This system has no alternative as there can exist only one maximum.

Thanks for clarification. This really highlights to the value of our engine. 

My thought process was based on the assumption that the engine calculate the maximum amount to exchange on given AMM which accounts to less than 5% change in price. Therefore I presumed exchanging x amount of tokens on two different exchanged would give same value, providing the amount is small (<5% change in price on both AMMs). 

The assumption is wrong.
Exchanging 10,000 Dpands to Algo does not return same value **even when exchange amount accounts to less than 5% change in price**
- Tinyman 3.053787 Algo
- Algofi 2.956357 Algo

It is clear to me now why our engine provide Tinyman as first choice. 

Great work really, I am fascinated by the degree of perfection ASA Stats has.

**Reply by ipaleka** [reply]
Thank you mate for another entry in your collection of clear, concise, and straightforward comments that leaves no open questions after reading. And of course, thanks for the compliments. :)

**Comment by ipaleka** [comment]
by kerrilija in Discord

- - - - - 

No Tinyman swap quote for Algodex token?

- - - - -

by ipaleka

Engine gives the maximum, Tinyman obviously doesn't have relevant enough price. https://github.com/asastats/channel/discussions/110#discussioncomment-2830141 

- - - - -

by kerrilija

I understand, but I'm still interested in the Tinyman price? I can see all the prices for other DEXes.

- - - - -

That is explained [here in this thread](https://github.com/asastats/channel/discussions/110#discussioncomment-2830141). 

> The actual price row exists solely for purpose of identifying imperfections of the system used. Right now the input parameters for our engine are (2, 2, 10) (let's skip their meaning for now). If we change those parameters to (0.01, 0.01, 90) the engine would make around 1000 times more calculations per bundle/address and all the AMMs would get exactly the same price (up to the ninth decimal), let's say it would become:
> 
> `0.39500107 ALGO/comand`
>
> This system has no alternative as there can exist only one maximum.

Let me explain that differently:

We do not present a price on AMM, we present the value you can get for your amount and then we divide that value with that amount to justify such a social construct as the price on AMM is.

ASA Stats shows you the **value** of your account and our engine uses all the available sources on the blockchain to get you the **real value** of your account.

**Reply by ipaleka** [reply]
by kerrilija

That total or now maximum value is actually a cornerstone of ASA Stats. But I believe a lot of people use ASA Stats to check prices of all the ASA they are interested in. I agree that maximum value should be the primary interest, or to say show what DEX offers the best price, but the price per DEX is also an interesting feature.

**Reply by ipaleka** [reply]
If you, or someone else, manage to find a proper way to implement such a feature, we'd add it to dependable features https://github.com/asastats/channel/wiki/FeaturesDependentOnUserWidgets.

**Reply by ipaleka** [reply]
You may use this thread to contemplate design or create a new discussion thread especially for the purpose.

**Reply by kerrilija** [reply]
How about an option (user preference)?
First option: prioritized DEXes that provide maximum value.
Second option: showing swap quotes from all DEXes(this should not affect total algo)

**Reply by ipaleka** [reply]
> First option: prioritized DEXes that provide maximum value.

That's oxymoron then. The option can only be a) maximum value b) don't use the following DEXes

> Second option: showing swap quotes from all DEXes(this should not affect total algo)

Basically, we should have add another system that isn't based on maximum value.

We'd need a wider analysis do we need that for sure and how exactly it should be marketed and presented in UI? Why are we adding it? Who are the users that prefer to have information about preferred DEX value instead of maximum?
